### ISSUE TITLE
native IDT in KVM? (FPU performance is bad)

### ISSUE THREAD

**Issue by stefand** [stsp/dosemu2#982]
Here's more FPU fun. Feel free to ignore this bug, it is probably not relevant in the real world.

http://www.conradshome.com/win31/archive/softlib/winfloat.exe contains a small FPU benchmark in floatapp.exe. It reports the following run times:

dosemu2 with _mathco=off or hide87: 6500-7000ms
dosemu2 with HW FPU: ~3500 ms
wine: 5ms

All 3 are on my i7 macbook pro (cpuinfo-grey.txt from bug #976). For comparison, real Windows 7 32 bit on a 10 years older machine (Core 2 Duo) runs the benchmark in 15ms. I have no clue how fast it would be on a Pentium 1 from 1995.

**To Reproduce**
Download and unzip http://www.conradshome.com/win31/archive/softlib/winfloat.exe, run floatapp.exe in Windows 3.x, click "benchmark" in the menu.

**Attach the log**
A -D+9M log is attached. Logging DPMI calls slows down things a lot more, so I guess the performance comes from entering and leaving the DPMI guest.

[boot.log.gz](https://github.com/stsp/dosemu2/files/4027016/boot.log.gz)

FloatApp also includes an exception test that reproduces bug 976 with HW FPU, but works OK with _mathco=off. Wine apparently can't handle Win16 FPU exceptions at all and just crashes in the 64 bit ntdll exception handler.

**Comment by stsp** [comment]
Try `speed 0` when starting an app.
Also the problem may be a writes to
0xf0/0xf1 FPU ports. See if `fpu_io_write()`
activates.

**Comment by stefand** [comment]
no change from speed 0

I put an fprintf line into fpu_io_write, it doesn't get triggered.

**Comment by stsp** [comment]
Could you try bot vm86 and DPMI to kvm?

**Comment by stefand** [comment]
Windows doesn't start with dpmi=kvm here:

ERROR: KVM: exit_reason = 0x6
ERROR: leavedos called recursively, forgetting the graceful exit!
removing stale files:

A case for another bug I guess :-)

**Comment by stsp** [comment]
Yes.

**Comment by stefand** [comment]
For reference, with the now-fixed KVM dpmi mode the benchmark needs about 2400ms with HW FPU and 5000ms with the FPU hidden. I am surprised it beats "native".

**Comment by stsp** [comment]
KVM is supposed to be faster than native
because this signal stuff is expensive, and
we need to do a few more syscalls from a
sighandler just to get things ready.

```
$ grep "int 0x3d" boot.log |wc -l
1048563
$ grep "to dosemu" boot.log |wc -l
1151967
```

So all the time it was calling `int 0x3d`.
How does wine speed that up?

**Comment by stsp** [comment]
Of course it is possible that int3d is
called by win87em.dll itself, which wine
does not do.

**Comment by stefand** [comment]
Ya, it is quite likely that the int comes from a win16 DLL that Wine handles at a higher level. Wine also implements the fpu emulation ints though: https://github.com/alexhenrie/wine/commits/master/dlls/krnl386.exe16/fpu.c

Looks like it is patching them out of the loaded code image and replaces them with actual fpu instructions. I am a bit puzzled here. My previous mental model of win87em fpu emulation did not involve special instructions in the program binary.

**Comment by stsp** [comment]
Bad URL on your side.
The good one is:
https://github.com/alexhenrie/wine/blob/master/dlls/krnl386.exe16/fpu.c

So I think this can be closed.

**Comment by stsp** [comment]
> Looks like it is patching them out

Good performance trick.
Why does it patch them with 1 byte when
intXX is 2 bytes?

**Comment by stefand** [comment]
Erm, I should even have used the official wine repo rather than a random github fork.

How does dosemu handle them? I'll see tomorrow if the replacement code triggers and what the performance of native win87em emulation mode is with hide87 on a real windows.

The disassembler I used to understand the netscape libs might have translated those int opcodes to fpu mnemonics

**Comment by stsp** [comment]
> How does dosemu handle them?

By just passing the control to the handler
registered by win87em.dll.

```
void WINAPI DOSVM_Int3dHandler(CONTEXT *context)
{
    TRACE("Int 0x3d called-- Standalone FWAIT\n");
    FPU_ModifyCode(context, 0x90);
}
```

0x90 is NOP. But it is 1 byte, and `int 3d` is `cd 3d`.
So what will we end with, `90 3d`?

**Comment by stefand** [comment]
It is patching it with 0x9b 0x90 aka fwait nop. The first byte (0xcd) turns into 0x9b unconditionally.

I would expect native win87em to do the same hotpatching. I wonder why it does not, and why wine implements it in krnl386 and not win87em.

https://reverseengineering.stackexchange.com/questions/12272/what-is-the-protocol-for-x87-floating-point-emulation-in-ms-dos

Speculates about those ints, in the context of a dos program. Might the replacement of those be a dos function rather than a windows/win87em one?

**Comment by stsp** [comment]
Yes: Borland was always putting the windows
compatibility layers in their products. 32rtm have
them all.

**Comment by stefand** [comment]
More factoids: floatapp.exe benchmark on real Windows 7 32 bit with hide87 is ~2200ms, about a factor of 150 compared to the 15ms with HW FPU. That's more likely to get me to buy an 80387 than a factor 2 speedup :-)

I still don't have hard evidence which code should be doing the patching (dos? dosx? krnl386? win87em?)

**Comment by stsp** [comment]
Why dont you think that its special
to wine? Might be a speed-up trick.

**Comment by stefand** [comment]
It is not special to Wine. I am just thinking out loud if Wine implements the trick in the correct place and, by extension, if dosemu2 should implement it in the handlers for those FPU emulation interrupts.

**Comment by stsp** [comment]
Why do you think its not special
to wine?

**Comment by stefand** [comment]
For one, Wine isn't magically faster than a real direct-on-hw Windows.

Apparently compilers from the 90's use those ints, so apparently some design work went into them, and Wine certainly wasn't a goal for those compiler devs. I doubt that Wine devs saw some ingenious speedup tricks 10 years after no-FPU CPUs were a thing.

I saw some more clues in dlls/krnl386.exe16/ne_segment.c:

```
        case NE_RELTYPE_OSFIXUP:
            /* Relocation type 7:
             *
             *    These appear to be used as fixups for the Windows
             * floating point emulator.  Let's just ignore them and
             * try to use the hardware floating point.  Linux should
             * successfully emulate the coprocessor if it doesn't
             * exist.
             */
            TRACE("%d: TYPE %d, OFFSET %04x, TARGET %04x %04x %s\n",
                  i + 1, rep->relocation_type, rep->offset,
                  rep->target1, rep->target2,
                  NE_GetRelocAddrName( rep->address_type, additive ) );

```

And that TRACE line gets triggered a lot. Looks to me like the binaries are supposed to contain proper FPU instructions, together with "relocation" information that causes Windows to replace them with ints (unconditionally? When no FPU is present?). That's why those ints don't occur on Wine - it doesn't bother to create them.

The question is, does Windows 3.11 in dosemu? If so, is the creation of the int instruction the bug, or the lack of translating them back later? (I can't find a NE loader in dosemu nor fdpp, so I guess it is Windows doing the loading)

**Comment by stsp** [comment]
There are many other ways of real
windows being fast in that case. All
you need is to set up the real IDT
entry for the handlers. dosemu/wine
have only the IDT emulation, and
that is slow, not intXX is slow.

Of course dosemu does not
natively load ne, windows does.

**Comment by stefand** [comment]
I doubt that even ideal int handling gets anywhere close to running a native FPU instruction. I don't have proof though :-)

To continue some random speculations: We might be able to abuse these relocation-introduced ints to chip away at bug 976. If fxsave isn't actually fxsave but an int that win87em handles we can patch up the saved FPU info before returning to the app. It won't fix problems where win87em calls fxsave for its own internal use because win87em doesn't have reloc info for patching up its own FPU instructions. But win87em.dll is small and replacable.

Again, feel free to ignore, there are probably more important things to fix. I might have fun with it in Wine some day, but Wine's 16 bit FPU code has its own share of troubles :-) .

**Comment by stsp** [comment]
intXX that does't switch protection
ring, is slow only by doing a couple
of stack pushes, like far call does
plus one more push for flags, and
also an IDT and LDT look-ups.
Its not a big deal in terms of 90s
programming. I would really be
surprised if both borland and ms
did the patching.

There is no such fxsave.
There is an fstenv in the int75
handler on win87em.dll that
looks up the address of the faulted
fpu instruction and does something
with it.

**Comment by stsp** [comment]
Well, we can actually use native IDT
under KVM to make this blazingly fast.
But its not a top priority for sure.

**Comment by jschwartzenberg** [comment]
I suppose QEMU would give you native IDT with KVM? It could be checked to see what performance would be like.

**Comment by stsp** [comment]
What will you run on qemu?
We already know that on native windows
it is fast, so under qemu it will remain so.
We already know that wine patches the
instructions, so under qemu it will remain
fast. There is nothing to check, I am afraid.

**Comment by jschwartzenberg** [comment]
It doesn't seem bad there's nothing to check :)

I think it would be good to have an overview of all the differences between running FDPP/FreeDOS inside QEMU+KVM versus DOSEMU2+KVM. For simple users this is not as clear as it is for developers.

**Comment by stefand** [comment]
What's worth checking in qemu is if Windows, running in its entirely native mode (Dos 6.22, its own dosx.exe, ...) patches the opcodes like Wine does. Doing that is on my todo list, but my day job called me back, so it'll take some time.

I know that Windows already patches the FPU opcodes to ints because the ints do not exist in the .exe file on disk. I suspect this is wrong and should not happen with a real FPU in place. But if it is supposed to patch them in the wrong direction on NE module load, I'd expect it to un-patch them on execution. Still, no proof without attaching a debugger and looking at the .exe code in memory after loading.

**Comment by stefand** [comment]
Actually, the correct way would be to write a test .exe that contains an FPU instruction and checks its own code after loading, but before executing that instruction and after executing that instruction. No need to mess with debuggers and it makes a nice regression test.

**Comment by stsp** [comment]
> I know that Windows already patches the FPU opcodes to ints because the ints
> do not exist in the .exe file on disk.

But when I was debugging netscape, I've seen
FPU instructions. And in fact, there would be no
FPU fault if all FPU instructions are patched to int.

**Comment by stefand** [comment]
That'd depend on the exe file having NE_RELTYPE_OSFIXUP 'relocation' instructions for the FPU instruction(s) you were looking at. PR1640.DLL seems to have plenty of them, but not for any offset 8c2f (or nearby), which is where bug 976's FPU fault happens. (Checked with that TRACE line in Wine I pointed to earlier).

It could also be a case of patch fpu -> int on load, run as int once, patch back to fpu, re-execute directly or a 2nd time later on and the 2nd execution then raises the FPU exception. An int 3x at the same location earlier on might give away something like that.

It also depends on the NE_RELTYPE_OSFIXUP comment in Wine being correct. floatapp.exe and floatdll.dll don't contain any int3x instructions. I checked the bytecode to make sure it wasn't the disassembler being 'smart' and translating them. Yet those ints come from somewhere.

Bottom line, we know about as much as Jon Snow...

**Comment by stsp** [comment]
Such things are very easy to check.
With dosemu2 you can use gdb to set the
HW watch-point to the address of the
instruction. You will see all the transformations
that are done to that address.
Of course you will say that dosemu2 is
not trust-worthy and the windows should be
run on qemu instead. But if not - its very
simple to verify. qemu also supports gdb
debugging, but I think there will be problems
related to paging etc to have the persistent
address. But under dosemu2 there are no
problems at all.

**Comment by stefand** [comment]
From boot.log.gz attached to https://github.com/stsp/dosemu2/issues/976#issuecomment-569345296 line 843280.

```
DPMI: cs=0x1387, ss=0xecf, ds=0x138f, es=0x138f, fs=0, gs=0 ip=0x8c1f
DPMI: int 0x003d, AX=0x138f
DPMI: int 0x3d
DPMI: call inthandler 0x3d(0x138f) at 0xf0f:0x00086e
		ret=0x1387:0x008c21
DPMI: cs=0xf0f, ss=0xecf, ds=0x138f, es=0x138f, fs=0, gs=0 ip=0x86e
DPMI: switch to dpmi
DPMI: Return to client at 0f0f:0000086e, Stack 0xecf:0x0000e03e, flags=0x93292
DPMI: switch to dosemu
DPMI: Return to dosemu at 1387:00008c33, Stack 0xecf:0x0000e044, flags=0x93202
DPMI: run_pm_int(0x75) called, in_dpmi_pm=0x01
DPMI: Switching to locked stack
DPMI: Calling protected mode handler for int 0x75
```

There is an int 0x3d right before the FPU exception. I am trying to find the line in the disassembly again, but somehow my cs:offset calculations aren't working out. I don't see any int instructions in all of pr1640.dll though.

The part that's confusing though is that if I am reading this right then something must have patched the int back to an fpu instruction for netscape - something that doesn't seem to happen for floatapp.exe.

**Comment by stsp** [comment]
You can add `D_printf("%s", DPMI_show_state(scp));` to
dump everything, including the opcode sequence.

**Comment by stefand** [comment]
Ok, I think I mis-identified the DLLs last time. I think the code is coming from js1640.dll, not pr1640.dll. Now offsets in %ip and disassembly match up nicely. Some relocs identified (and ignored) by Wine:

```
009c:err:fixup:apply_relocations 115: TYPE 7, OFFSET 8c1b, TARGET 0005 0000 OFFSET16 add name JS1640
009c:err:fixup:apply_relocations 116: TYPE 7, OFFSET 8c1f, TARGET 0006 0000 OFFSET16 add name JS1640
009c:err:fixup:apply_relocations 117: TYPE 7, OFFSET 8c2a, TARGET 0005 0000 OFFSET16 add name JS1640
009c:err:fixup:apply_relocations 118: TYPE 7, OFFSET 8c2e, TARGET 0005 0000 OFFSET16 add name JS1640
009c:err:fixup:apply_relocations 119: TYPE 7, OFFSET 8c32, TARGET 0005 0000 OFFSET16 add name JS1640
009c:err:fixup:apply_relocations 120: TYPE 7, OFFSET 8da7, TARGET 0005 0000 OFFSET16 add name JS1640

```

The one at offset 0x8c32 is our special friend from #976 

At offset 0x8c1f an int 0x3d shows up repeatedly in the -D+9M log, but not the others. It has TARGET=6, and a 0x90 NOP in the DLL file.

Some ints do get patched in, although it seems to be somewhat selective.

**Comment by stsp** [comment]
So do that relocations patch in "int OR FPU",
depending on an FPU presence? Or do they
always patch in int?
That still gives us no evidence of the wine-alike
int-to-FPU patching, right?

**Comment by stefand** [comment]
> So do that relocations patch in "int OR FPU",
> depending on an FPU presence? Or do they
> always patch in int?
I don't know.

> That still gives us no evidence of the wine-alike
> int-to-FPU patching, right?
Correct. It just shows the loader patches in an int. We still don't know if this is correct or not, and if the loader is behaving correctly, whether it should get patched back out or not on execution.

The instances of target 6 relocs I see follow fnstcw instructions. Is there any reason why those are of special interest in the presence of an FPU? E.g. could windows try to fix up some flags if it is running on an older CPU/FPU?

**Comment by stsp** [comment]
> Is there any reason why those are of special interest in the presence of an FPU?
> E.g. could windows try to fix up some flags if it is running on an older CPU/FPU?

I think it just does `fwait` after them, which is `int 3d`.

> The instances of target 6 relocs I see follow fnstcw instructions.

Are they in a binary?

**Comment by stefand** [comment]
> > The instances of target 6 relocs I see follow fnstcw instructions.
> 
> Are they in a binary?

Yeah, they are part of the NE binary.

I looked at the running floatapp.exe in a windows debugger inside qemu and the int 0x3d's are there as well. Performance is sane (~275ms with qemu HW fpu, ~20000 with hide87), so I think we now have certainty that your diagnosis - ints are too slow - is correct. Sorry, I was too lazy to hand-craft a NE module with those relocations that later on self-examines itself..

There's still the remote possibility that the debugger doesn't pick up runtime code changes, but I think this is now bordering on a conspiracy theory. It did see the load time nop -> int 0x3d changes, so I assume it would also pick up int 0x3d -> fwait changes if they should happen.

Unrelatedly, qemu's mouse support sucks. Back to Wine's ddraw tests now...

**Comment by stsp** [comment]
> Unrelatedly, qemu's mouse support sucks.

Really?
And I have just recently noticed that it seems
to follow the host pointer even w/o grabbing input
(in windows-10 as guest), so I thought its a major
improvement. Possibly there is no proper driver
for win31, and they forgot to add the needed int15
code to seabios?

> certainty that your diagnosis - ints are too slow - is correct.

And that wine does a speed-up hack, right?

**Comment by stefand** [comment]
> > Unrelatedly, qemu's mouse support sucks.
> 
> Really?

Ya, if you go in and out of mouse grab mode too much it'll randomly start misinterpreting moves as clicks, etc. And without mouse grab mode there's no input

> And I have just recently noticed that it seems
> to follow the host pointer even w/o grabbing input
> (in windows-10 as guest),

I think I saw the same in VirtualBox and I think I saw the mouse labelled as something like "VMWare guest mouse" or so, so I guess Windows just brings something like your int33mouse.

> Possibly there is no proper driver
> for win31, and they forgot to add the needed int15
> code to seabios?

Well, mouse support works, it's just buggy. I guess nobody used Win3x in ages.

> > certainty that your diagnosis - ints are too slow - is correct.
> 
> And that wine does a speed-up hack, right?

More like a missing feature. The loader doesn't honor the original NE reloc instructions to add those int 0x3d opcodes, so Wine just executes the original fwait nop from the executable image.

But yeah, the later handler that un-patches any int 0x3d's that might occur could be considered a hack. But if I understand it right that's only a default handler that would be replaced by a proper win87em.dll implementation. So call it a missing feature too :-D

**Comment by stsp** [comment]
> More like a missing feature. The loader doesn't honor the original NE reloc instructions
> to add those int 0x3d opcodes, so Wine just executes the original fwait nop from the
> executable image.

But earlier you said you see NOPs in some image js1640.dll?
`It has TARGET=6, and a 0x90 NOP in the DLL file.`

> But if I understand it right that's only a default handler that would be replaced by a proper
> win87em.dll implementation.

But how it wouldn't be slow then?

**Comment by stefand** [comment]
What I meant with that is that the file on disk contains "0x90 0x9b" . Wine loads and executes it as-is. Windows replaces it with int 0x3b. Wine is fast because it never puts any ints into the code. It could also run into some behavior bugs due to that if win87em should do something in the interrupt handler.

E.g. Wine's support for Win16 FPU exceptions is nonexistent. Netscape crashes on the FPU exception when our ntdll looks for a Win32 exception handler in the TEB, finds only UnhandledExceptionFilter() at the bottom of the stack and invokes it. I am not sure how much other Win16 exception handling is there. I see some fragments of what looks like DPMI fault handling but I doubt it is complete.

**Comment by stsp** [comment]
> What I meant with that is that the file on disk contains "0x90 0x9b" .

But you said `0x90 NOP`, did you mistype a `0x90 fwait`?
Then that would make sense. :)
And the patching code is then unused?

> Netscape crashes on the FPU exception when our ntdll looks for a Win32
> exception handler in the TEB

Can't wine instead fire the IRQ13 as dosemu2 does?

**Comment by stefand** [comment]
> But you said `0x90 NOP`, did you mistype a `0x90 fwait`?
> Then that would make sense. :)

No, that was more to say that the nop instruction is a 0x90. I came across different nops, mostly longer, e.g. 0x8b 0xff. Look up Win32 hotpatching for details on that one. I got burned by that stuff in the past...

> And the patching code is then unused?

The int 0x3x -> FPU opcode patching? Yeah, for this particular app it is unused.

> > Netscape crashes on the FPU exception when our ntdll looks for a Win32
> > exception handler in the TEB
> 
> Can't wine instead fire the IRQ13 as dosemu2 does?

It presumably should. That, or int 0x75, or directly invoke whatever has been set with http://www.delorie.com/djgpp/doc/dpmi/api/310212.html . I have to study the code to figure out what would be the right approach. Wine has infrastructure for setting custom int callback functions.

**Comment by stsp** [comment]
> The int 0x3x -> FPU opcode patching? Yeah, for this particular app it is unused.

If you say the image initially contains the FPU ops
and wine never translates them to intXX, then I
suppose the runtime-patching code is always unused.

> It presumably should. That, or int 0x75

No PIC emulation in wine I suppose?

> directly invoke whatever has been set with

You probably meant this one:
http://www.delorie.com/djgpp/doc/dpmi/api/310203.html

> I have to study the code to figure out what would be the right approach.

Would be interesting to find that out.
Eg if win9x used to set CR0.NE and then called
int0x75 for win16 apps, then it would be very
tempting to call int75 directly under dosemu2
as well. I am just worried about PIC emulation,
who knows if the vector is moved...

**Comment by stsp** [comment]
Ah wait, doesn't calling int75 directly, accounts for an EOI?

**Comment by stefand** [comment]
> > The int 0x3x -> FPU opcode patching? Yeah, for this particular app it is unused.
> 
> If you say the image initially contains the FPU ops
> and wine never translates them to intXX, then I
> suppose the runtime-patching code is always unused.

There might be other images generated by other compilers that have the ints out of the box.

> > It presumably should. That, or int 0x75
> 
> No PIC emulation in wine I suppose?

Not sure; I'd have to read the code. I would be very surprised if we have HW level emulation. It looks like we just get an invalid instruction for SW interrupts and invoke the registered DPMI handler: dlls/krnl386.exe/instr.c:

```
        case 0xcd: /* int <XX> */
            context->Eip += prefixlen + 2;
            if (DOSVM_EmulateInterruptPM( context, instr[1] )) return ExceptionContinueExecution;
            context->Eip -= prefixlen + 2;  /* restore eip */
            break;  /* Unable to emulate it */

```
 
> You probably meant this one:
> http://www.delorie.com/djgpp/doc/dpmi/api/310203.html

I was taking the "don't use this, go for the DPMI 1.0 one" at face value :-)

> Would be interesting to find that out.
> Eg if win9x used to set CR0.NE and then called
> int0x75 for win16 apps, then it would be very
> tempting to call int75 directly under dosemu2
> as well. I am just worried about PIC emulation,
> who knows if the vector is moved...

```
    case 0x0203:  /* Set Processor Exception Handler Vector */
         FIXME( "Set Processor Exception Handler Vector (0x%02x)\n",
                BL_reg(context) );
         break;

```

I am a bit puzzled though, I did not see a FIXME about stuff like that when I ran floatapp.exe in Wine and tried to trigger an unmasked FPU exception.

From skimming the code I'd expect anything Win16 in Wine that receives a SIGSEGV to die unceremoniously. I only see the exception handler for invalid instructions. There's also a handler that intercepts access to the lower 1 MB of memory and puts fake BIOS info in place. dosvm.c has some interrupt handler magic, but afaics it just puts some 0xcd instructions in place to trigger the invalid instruction handler.

We've also outsourced the business of running actual real mode DOS executables to dosbox.

> Ah wait, doesn't calling int75 directly, accounts for an EOI?

No idea what this means :-( .

**Comment by stsp** [comment]
But since when intXX became an invalid opcode?
It should be the same SIGSEGV?

> intercepts access to the lower 1 MB of memory

How? Page-protects all low memory?

> We've also outsourced the business of running actual real mode DOS executables to dosbox.

Try fdpp. :)

> No idea what this means :-(

This is a procedure of notifying PIC (via IO ports)
about the IRQ completion. It then updates its
internal registers to accept another IRQs.

**Comment by stefand** [comment]
> But since when intXX became an invalid opcode?
> It should be the same SIGSEGV?

On the Linux side yes. ntdll will translate it to EXCEPTION_PRIV_INSTRUCTION. krnl386.exe16 talks Win32, not POSIX.

> > intercepts access to the lower 1 MB of memory
> 
> How? Page-protects all low memory?

It comes pre-protected in the default Linux and Windows VM setup. I just see there's some magic to handle situations where Linux forbids us to unprotect the low 64k address space, and we're looking if an application creates a segment descriptor with a base < 1MB.

> > We've also outsourced the business of running actual real mode DOS executables to dosbox.
> 
> Try fdpp. :)

We're also not in the business of emulating most hardware :-) . I see 8259 timers, parport in krnl386/ioports.c, and there's soundblaster, but no VGA and no PIC. parport uses ppdev, some code for you to look at. I can't even find COM port emulation, but maybe there's some magic in ntdll.

**Comment by stefand** [comment]
Oh, and no CPU emulator. The lack of vm86 is what motivated our switch to dosbox. It was way easier than integrate a CPU emulator. We also had VGA emulation at some point, but it was terrible and we removed it with the vm86 support.

**Comment by stsp** [comment]
> On the Linux side yes. ntdll will translate it to EXCEPTION_PRIV_INSTRUCTION.

You mean, if no handler was registered?

> I just see there's some magic to handle situations where Linux forbids us to
> unprotect the low 64k address space

You should try shadow LDT bases.
This is the most ingenious trick that I inherited
from dosemu1.

> We're also not in the business of emulating most hardware :-)

Yes but fdpp does not require those.
It uses bios interrupts to access the IO
hardware (com, lpt, video). And bios
interrupts can return "no such hardware".
I really think fdpp is very pluggable and
doesn't have too many requirements.
It can turn any project into a DOS emulator
rather quickly. :)

> Oh, and no CPU emulator. The lack of vm86

Yes, this is quite a problem indeed. :(
Of course with KVM you can get vm86 back,
but then its not portable.
Is there an integration with dosbox, like winoldap?

**Comment by stsp** [comment]
I checked that int75 handler does EOI (as it should)
to both PICs. Fortunately it does so before the
GPF exception occurs on bad selectors, or there
wouldn't be any way to cleanly recover.
So if calling int75 directly, one have to be prepared
to an attempt to terminate the interrupt that, from the
PIC's standpoint, was not started. But wine's win87em.dll
probably doesn't hook int75 at all, and so, no such
problem exist there?
I wonder what win9x's win87em.dll does.

**Comment by stsp** [comment]
I implemented the use of native IDT
(I hope... can't see any changes),
all still works.
Could you please test that FPU?

**Comment by stsp** [comment]
Wishful thoughts...
Forgot to change DPL.

**Comment by bartoldeman** [comment]
@stsp yes, I'm still alive

I read this thread with interest and hacked a bit myself, but your code is cleaner. However I think you're missing a few things:
* probably you should also exclude 0x70 to 0x77 (hardware ints). I think excluding 0x1c, 0x23, 0x24 is good too (just being cautious unless they prove ok), FWIW cwsdpmi excludes hw ints and 0x1c, 0x23.
* dpmi_init needs to call dpmi_set_interrupt_vector() too
* need to worry about dosdebug?

Here is my patch for reference:
https://gist.github.com/2beb51555a4d6ecaf56e02d2783c2403

I also timed it using a compilation of freecom using openwatcom with "speed 0" (incl. dosemu boot time):
```
native:
real    2m2.689s
user    1m5.841s
sys     0m55.986s
before IDT patch:
real    1m54.695s
user    1m3.068s
sys     0m50.807s
after patch:
real    1m38.569s
user    0m55.733s
sys     0m41.970s
Linux cross compilation:
real    0m4.920s
user    0m3.631s
sys     0m1.244s
```
I was also thinking that perhaps it would be slightly faster for the default int handlers to simply hook the interrupt and then call HLT instead of generating a GPF (by using DPL=3 instead of DPL=0).

**Comment by stsp** [comment]
> @stsp yes, I'm still alive

Good to know! :)

> I read this thread with interest and hacked a bit myself, but your code is cleaner.

Well, it doesn't work (if I also change DPL),
so I unpushed it. Btw, is there any reason
why saving regs by hands is done, rather
than via mmap'ed kvm region?

I don't see the immediate need to exclude
HW ints, as I suppose they usually do not
go via IDT in dosemu. PIC calls to DPMI
directly, which then uses its internal table
and cooks up the iret frame by hands.
As for 0x1c, 0x23, 0x24 - they usually
come from RM IIRC, which should also
be unaffected.
But, as my code doesn't work, all I say
here doesn't matter much. :) There is a
mistake somewhere in my understanding
or coding.

> dpmi_init needs to call dpmi_set_interrupt_vector() too

Indeed.

> need to worry about dosdebug?

Its mostly unaffected, as it usually steps
_over_ the intXX. If you use "ti" to trace into,
then it should simulate intXX by hands,
bypassing IDT.

> Here is my patch for reference:

Oh, cool! Does it work? :)
Why not a PR? Will test.

> I was also thinking that perhaps it would be slightly faster

Yes, indeed, would also make the patch smaller!

**Comment by stsp** [comment]
Ah, you mean bpint in dosdebug...
Indeed. :)

**Comment by stsp** [comment]
Your patch indeed works!
I applied it to the separate branch.
Have to go to sleep now.
@stefand could you please test
Bart's patch in branch kvm_fpu2?

**Comment by bartoldeman** [comment]
PR submitted.

If I change DPL the asm or C code needs to change a bit too since it no longer comes back with the same trapno and error code (it's a software int so just return cs:eip and flags pushed to the stack), so that's a bigger patch in the end. I hope to work on that later.

As for manually getting/setting registers that was the only thing possible at the time. I just found out about kvm_sync_regs and indeed it is a bit newer (2017) than my initial KVM patches (2015/2016), see https://patchwork.kernel.org/patch/9961983/.

**Comment by stsp** [comment]
Thanks!
It doesn't build though. :)

So the short-term KVM todos seems to be:
1. get rid of manual registers save/restore
2. get rid of this irq0x21 of 0xff
3. use DPL=3 for IDT even in default case

And a slightly more difficult one:
4. make it to work with simx86's page protection
that it puts on lowmem. This will require the
kvm_base patch, or maybe "map directly to
source" approach as you tried before.
After that, I'd be happy to enable KVM by
default.

**Comment by jschwartzenberg** [comment]
I don't know if IDT and LDT are related other than looking like similar abbreviations? I see there's also LDT R/W emulation. Would it make sense to have this native with KVM as well?

**Comment by stsp** [comment]
LDT write is used only by win31
and isn't frequent enough to be
perfarmance-critical. OTOH it is
more difficult to implement than
native IDT, because in case of IDT,
it is always updated with dpmi api,
just can be used natively. LDT is
already updated with dpmi api and
is used natively (even w/o kvm -
modify_ldt() syscall). Getting it both
updated and used natively, goes
far beyond of what we do for IDT.

**Comment by stsp** [comment]
@stefand Could you please check that its fixed?

**Comment by stefand** [comment]
Benchmark runtime decreased from ~3000 ms to ~160 ms. Nice!!!

**Comment by stefand** [comment]
Float emulation improved from ~6500 to ~450 ms. Now I can either rant that the the performance gain from hard float still way smaller than it is on native Windows / and qemu, or be happy that both modes are improved by a factor of 15-20 :-) .

**Comment by jschwartzenberg** [comment]
It would be good to figure out why QEMU is faster than DOSEMU2 with KVM I think.

**Comment by stsp** [comment]
Please attach the same log again,
and maybe we'll see what it does now.

**Comment by stefand** [comment]
> It would be good to figure out why QEMU is faster than DOSEMU2 with KVM I think.

It isn't, at least not in raw numbers. qemu with hw float was 265 ms, dosemu 160. With SW it is 20000 vs 450. Smaller is better in this case, as it is just a dumb timed loop doing nonsense float calculations.

A D+9M log is attached
[boot.log.gz](https://github.com/stsp/dosemu2/files/4051648/boot.log.gz)

**Comment by stsp** [comment]
But if it isn't worth, then should I look
into the log?

**Comment by stefand** [comment]
If you suspect that there are further optimization options it might be worth a look.

If I understand the patch right it shouldn't just affect those floating point ints, but also int instructions that call DOS APIs right? So is there potentially an across the board perf improvement from this?

**Comment by stsp** [comment]
The ones that call DOS API should
go via dpmi server to real mode anyway.
What can be speeded up, is only the
ints that have their PM handers that
do iret at the end (if they call to prev
handler - still slow). I think it will give
some improvement overall, but not
as big as it is here.

**Comment by bartoldeman** [comment]
Well the timings above show 15% improvement in file-system heavy performance.
What happens often is that a DPMI prog hooks an interrupt and then chains to the original. Before the patch (and always using "native") what you then get is:

DPMI -> int 21 -trap(GPF)> dosemu -> DPMI pushf call far orig handler -trap again via HLT instruction (GPF) -> dosemu (orig handler)

and now in KVM

DPMI -> int21 -goes directly to-> DPMI pushf call far orig handler -trap via HLT> dosemu

the shortcut happens quite frequently - in KVM it means it's only exited once instead of twice (if chained), or not at all, if taken over.

This is all quite analogous to the situation in VM86 where we also can tell that a software int goes straight to the procedure via the CPU instead of being trapped. With KVM that is done via VME, and with the old vm86() syscall it's emulated in the Linux kernel (it's still trapped but stays in kernel space).

**Comment by stsp** [comment]
But the question here is just how
frequently is int21 hooked in PM by
an app. To my memory this is not
usual.

**Comment by stsp** [comment]
I mean, you can find an apps where
that helps, but its not all apps, likely
not most apps.

**Comment by stsp** [comment]
> But the question here is just how
> frequently is int21 hooked in PM

OK, after re-checking - even those extenders
that rely on DPMI doing the int21 handling,
still hook int21 themselves too.
So yes, that is going to speed up everything. :)

**Comment by stefand** [comment]
This float test is probably an extreme case, as it does relatively little work itself, calls *lots* of ints, that themselves do relatively little work on their own. For a different perspective, here are some winbench 3.11 graphics test results. I don't know what exactly winbench does, but I guess it just calls gdi.exe and user.exe via regular 16 bit protected mode far calls, which then do some software graphics rendering and update the SVGA frame buffer. In the end it gives me a magic number, higher is better in this case

```
        native: 43317673 pixel/s
kvm idt emu   : 48022015 pixel/s
kvm idt native: 48143087 pixels/s
```

This is on the macbook pro. The difference between the old and new IDT is within the general ups and downs.

It has a disk benchmark too. I guess it degenerates into a syscall overhead benchmark on my hardware, with an SSD,16 gb of ram and a Linux kernel that does disk caching whether the benchmark wants it or not. The numbers are again an arbitrary score. The program does reads and writes of various sizes, the largest of which is 4096 byte.

```
        native: 5.6 MiB/s
kvm idt emu   : 7.2 MiB/s
kvm idt native: 7.1 MiB/s
```

Speculation follows:

So yeah, a difference between native and kvm can be seen, but the IDT change impact appears negative - although it's presumably just a general up and down. For comparison though, on the score screen it tries to sell me a Compaq 486/33m by telling me it scored 45 KiB/sec. (and no, it doesn't tell me what disk they used). So unless you are trying to host a Win16 database server I doubt it matters. Either your computer is waaay faster than old 1995 HW, or you'll have such a slow disk that the syscall overhead doesn't matter anyway.

More factoid: It gives me more fine grained data. 4kb sequential reads are ~63 MiB/s, whereas 200 byte random reads are ~2.3 MiB/s.

Linux native:
# hdparm -t /dev/sda

/dev/sda:
 Timing buffered disk reads: 1010 MB in  3.01 seconds = 336.06 MB/sec

It is supposed to give me ~550 MB/sec, but it is attached to USB 3, which is also supposed to be fast, but sucks in practice.

**Comment by stsp** [comment]
```
$ grep "Direct LDT" boot.log |wc -l
23321
 grep "int 0x" boot.log |wc -l
17412
```
So we have an almost even split between
remaining interrupts (the ones that now go
to real mode) and LDT writes.
In that scenario I suppose @jschwartzenberg 's
idea to allow native LDT writes may indeed
play the role.
I think the simplest and most efficient approach
here is to use PML. Bart, you have an experience
with PML. What efforts do you estimate would
require to allow native LDT writes with PML syncing?

**Comment by stsp** [comment]
I opened #988 for native LDT updates.

**Comment by stsp** [comment]
I made dpmi coopth to be created only once.
Could you please check if that is faster?

**Comment by jschwartzenberg** [comment]
So no native IDT in KVM?

**Comment by stsp** [comment]
> So no native IDT in KVM?

???

**Comment by jschwartzenberg** [comment]
Maybe I missed something, but the last commit is a revert.
